language/stdlib/modules/approved_payment.move

```diff
@@ -0,0 +1,95 @@
+// Module that allows a payee to approve payments with a cryptographic signature. The basic flow is:
+// (1) Payer sends `metadata` to the payee
+// (2) Payee signs `metadata` and sends 64 byte signature back to the payer
+// (3) Payer sends an approved payment to the payee by sending a transaction invoking `deposit`
+//     with payment data + signature. The transaction will abort if the signature check fails.
+// Note: approved payment's are an accounting convenience/a courtesy mechansim for the payee, *not*
+// a hurdle that must be cleared for all payments to the payee. In addition, approved payments do
+// not have replay protection.
+address 0x0:
+module ApprovedPayment {
+    use 0x0::Libra;
+    use 0x0::LibraAccount;
+    use 0x0::Signature;
+    use 0x0::Transaction;
+    use 0x0::Vector;
+
+    // A resource to be published under the payee's account
+    resource struct T {
+        // 32 byte EdDSA public key whose counterpart must be used to sign the payment data
+        public_key: vector<u8>,
+        // TODO: events?
+    }
+
+    // Deposit `coin` in `payee`'s account if the `signature` on the payment metadata matches the
+    // public key stored in the `approved_payment` resource
+    public fun deposit<Token>(
+        approved_payment: &T,
+        payee: address,
+        coin: Libra::T<Token>,
+        metadata: vector<u8>,
+        signature: vector<u8>
+    ) {
+        // Sanity check of signature validity
+        Transaction::assert(Vector::length(&signature) == 64, 9001); // TODO: proper error code
+        // Cryptographic check of signature validity
+        Transaction::assert(
+            Signature::ed25519_verify(
+                signature,
+                *&approved_payment.public_key,
+                copy metadata
+            ),
+            9002, // TODO: proper error code
+        );
+        LibraAccount::deposit_with_metadata<Token>(payee, coin, metadata)
+    }
+
+    // Wrapper of `deposit` that withdraw's from the sender's balance and uses the top-level
+    // `ApprovedPayment.T` resource under the payee account.
+    public fun deposit_to_payee<Token>(
+        payee: address,
+        amount: u64,
+        metadata: vector<u8>,
+        signature: vector<u8>
+    ) acquires T {
+        deposit<Token>(
+            borrow_global<T>(payee),
+            payee,
+            LibraAccount::withdraw_from_sender<Token>(amount),
+            metadata,
+            signature
+        )
+    }
+
+    // Rotate the key used to sign approved payments. This will invalidate any approved payments
+    // that are currently in flight
+    public fun rotate_key(approved_payment: &mut T, new_public_key: vector<u8>) {
```

Comment:
Out of curiosity, what is happening if I don't have a registered key and try to rotate it?

Reply:
`borrow_global<T>(a)` and `move_from<T>(a)` abort if there is no resource of type `T` published under address `a`. Similarly, `move_to_sender<T>` aborts if there is no resource of type `T` published under the sender's address. Basically, there's an implicit `exists` check in each case that will trigger an abort. 

We could choose to do an explicit `exists` check here if we wanted to abort with a more constructive error code. The cases above will abort with something fairly generic like `MISSING_DATA`. The circumstances under which you want to add such error codes is a topic of active debate in our team--it's a gas/code size vs error code clarity tradeoff.